keylime/registrar_common.py

```diff
@@ -291,20 +292,84 @@ def do_POST(self) -> None:
+                if not cert_utils.verify_cert(idevid_cert_509, config.get("tenant", "tpm_cert_store"), "IDevID"):
+                    web_util.echo_json_response(self, 400, "Error: IDevID certificate could not be verified")
+                    return
+                logger.debug("IDevID cert verified.")
+                if not cert_utils.verify_cert(iak_cert_509, config.get("tenant", "tpm_cert_store"), "IAK"):
+                    web_util.echo_json_response(self, 400, "Error: IAK certificate could not be verified")
+                    return
+                logger.debug("IAK cert verified.")
+
+                #TPM cert checks
+                if not cert_utils.check_tpm_origin(idevid_cert_509, "IDevID"):
+                    web_util.echo_json_response(self, 400, "Error: IDevID certificate might not be from a TPM")
+                    return
+                if not cert_utils.check_tpm_origin(iak_cert_509, "IAK"):
+                    web_util.echo_json_response(self, 400, "Error: IAK certificate might not be from a TPM")
+                    return
+                logger.debug("IDevID and IAK seem to come from a TPM")
+
+                # NEEDS CRYPTO >= 38.0.0
+                # Check that the agent-generated keys match the certificates
+                iakpub_cert = iak_cert_509.public_key()
+                iakpub_reg = tpm2_objects.pubkey_from_tpm2b_public(base64.b64decode(iak_tpm))
+                if iakpub_cert.public_bytes(
+                    serialization.Encoding.PEM, serialization.PublicFormat.SubjectPublicKeyInfo
+                ) == iakpub_reg.public_bytes(
+                    serialization.Encoding.PEM, serialization.PublicFormat.SubjectPublicKeyInfo
+                ):
+                    logger.info("IAK cert and key match")
+                else:
+                    logger.warning("Generated IAK does not match certificate %s", agent_id)
+                    web_util.echo_json_response(
+                        self, 400, "Error: Generated IAK does not match certificate. Check selected template."
+                    )
+                    return
+
+                idevidpub_cert = idevid_cert_509.public_key()
+                idevidpub_reg = tpm2_objects.pubkey_from_tpm2b_public(base64.b64decode(idevid_tpm))
+                if idevidpub_cert.public_bytes(
+                    serialization.Encoding.PEM, serialization.PublicFormat.SubjectPublicKeyInfo
+                ) == idevidpub_reg.public_bytes(
+                    serialization.Encoding.PEM, serialization.PublicFormat.SubjectPublicKeyInfo
+                ):
+                    logger.info("IDevID cert and key match")
+                else:
+                    logger.warning("Generated IDevID does not match certificate %s", agent_id)
+                    web_util.echo_json_response(
+                        self, 400, "Error: Generated IDevID does not match certificate. Check selected template."
+                    )
+                    return
```

Comment:
This do_POST is already so long. If you could put all these IAK/IDevID checks (starting at `# IDevID and IAK cert checks`) into a function returning bool that would be helpful. If it returns FALSE, just do `return` then.

Reply:
Better: have it return an error string and if there is one send it via web_util.echo_json_reponse()...

Reply:
I have moved the cert checks into a new function in cert_utils, and also rearranged so all idevid stuff is within one block